Issue #3675: Replace Scope with AccessModifier in ParameterNameCheck to avoid wrong scopes comparison

MEZk opened the pull request:
#3675 

1) Fixed problem with scope comparison. 'Scope' is incorrect term for this check. 
2) Added new enum 'AccessModifiers'. I have a concern: should the new enum be a part of Checkstyle's API now?
3) Properties 'scope' and 'excludeScope' became 'accessModifiers' and 'excludeAccessModifiers'. I don't think that this check needs 'excludeAccessModifiers' option anymore since now the check validates only methods which access modifiers are specified in 'accessModifiers'.
4) Added new util method 'getAccessModifierFromModifiersToken'. I'm not sure whether it should be in CheckUtils class or in AccessModifiers enum.
5) Fixed UTs and ITs. Added new one for 'getAccessModifierFromModifiersToken'.
5) Fixed UTs which validate xml-configs and xdoc.
6) Updated javadoc and xdoc. 
7) Fixed ParameterName to avoid unclear logic.

@romani, @rnveach 
Please check and let me know what do you think about 2,3 and 4.

rnveach commented:
@MEZk Issue is about ParameterName and scope.
https://github.com/checkstyle/checkstyle/pull/3722/files#diff-c55cd603ef6597463971db3f33f4f4c8
Why is there alot of changes outside this scope in google_checks? You are removing checks, adding new tokens, removing ids, etc.
Even UT on google_checks is complaining.
> config_whitespace.xml section 'ParenPad' should be in google_checks.xml or not reference 'Google Style'

romani commented:
I do not remember that I gave you permission to extend API now , it was planned to be done in checkstyle8. Please make it as non API class.

MEZk commented:
@rnveach 
Sorry, I cherry-piked final commit from my local test branch to PR branch incorrectly. Fixed.

MEZk commented:
@romani 
> I do not remember that I gave you permission to extend API now , it was planned to be done in checkstyle8. Please make it as non API class.

It will require 
1) making 'setAccessModifiers' receive an array of strings as a parameter.
2) implementing of converter from array of strings which represent access modifiers into array of elements of AccessModifiers enum inside ParameterNameCheck.
3) implementing of guard-conditions to check whether user passed correct values into setAccessModifiers
4) reimplementing of converters in future inside AutomicBean

Is it what we need now for the check? Or did you mean that AccessModifier should be outside api package but still accessible from AutomicBean while converting values received from xml-config?

rnveach commented:
> that AccessModifier should be outside api package but still be accessible from AutomicBean while converting values received from xml-config

Import control will not like that we are including an import from outside the API package for AutomaticBean. This is another version 8 change.

MEZk commented:
@rnveach 
> Import control will not like that we are including an import from outside the API package for AutomaticBean. 

Right, but I don't like the approach with string converting inside ParameterNameCheck either.

MEZk commented:
@romani 
Please, help us to decide.

romani commented:
Ideally I would prefer to stay with String and inside-Check conversion and validation, but ...

> Import control will not like that we are including an import from outside the API package for AutomaticBean.

We could do exception in rules only because checkstyle8 is very close, and implementation is done and MEZk is very focused contributor.
If @MEZk , will create separate issue to explain till what time it stay in code and will help us to resolve it in future, then I am ok keep code and do temporal allowance in import control config.

@MEZk , please take a look at shippable failure, pitest failures are not expected :).

rnveach commented:
@MEZk ParamaterNameCheck lost 3 points in mutation.
````
111 		1. removed call to java/util/Arrays::sort → SURVIVED
121 		1. removed call to java/util/Arrays::sort → SURVIVED
186 		1. changed conditional boundary → SURVIVED
````

MEZk commented:
@romani 
> please take a look at shippable failure, pitest failures are not expected

Fixed.

@rnveach 
> ParamaterNameCheck lost 3 points in mutation

Fixed. Read a few articles about mutation testing. Impressive style of testing. I think I'll enforce it at work)

@romani, @rnveach,
Please, address your thoughts about point number 3.

@romani 
> If @MEZk , will create separate issue to explain till what time it stay in code and will help us to resolve it in future, then I am ok keep code and do temporal allowance in import control config.

Yes, I'll do this.

rnveach commented:
@MEZk I agree there is no point for `exclude` now with the code as you wrote it, but your PR is completely removing the original properties in a minor release. Users won't be able to use the same config between this release. I believe we can't remove the properties and have to keep working with them until CS 8 where we break backwards compatibility.
This issue is just to fix the behavior of the check, unless @romani wishes to change that.

MEZk commented:
@rnveach 
Do you mean that we should stay with 'scope' and 'excludeScope' as properties names, but fix logic as described in the issue?

romani commented:
@MEZk , 

> implementing of guard-conditions to check whether user passed correct values into setAccessModifiers

yes, please add validation and throw illegalArgumentException.

>  I believe we can't remove the original properties and have to keep working with them until CS 8 where we break backwards compatibility.

It is better to remove design problems as soon as possible, to stop spread if its usage ASAP.
As we come to CSv8 , we will do such a damage that non of plugins will easily catch it and release new version, so updates in CS8 will be used by user in months after release of first CSv8. 
Only in this case (of pre v8) I am ok to change property names in minor release.
This properties will not be used by google style, as Google confirmed that skip all validation for length of arguments and allow any.

romani commented on src/main/resources/google_checks.xml:
```diff
@@ -108,16 +108,9 @@
              value="Member name ''{0}'' must match pattern ''{1}''."/>
         </module>
         <module name="ParameterName">
-            <property name="id" value="ParameterNameNonPublic"/>
```
if now need - remove it.

MEZk commented:
@romani 

> yes, please add validation and throw illegalArgumentException.

I have already implemented guard-condition. I cannot decide whether getAccessModifierFromModifiersToken should be in ChecksUtils or in AccessModifier. @rnveach , @romani, can you help me?)

> It is better to remove design problems as soon as possible, to stop spread if its usage ASAP.

Done. Properties were replaced with new ones. After a merge of this PR I will make a note at #3511 in order not to forget that we need to move AccessModifier to api package.

> I am ok keep code and do temporal allowance in import control config.

Please, check import-control.xml

@rnveach 

I fixed the problem in google_checks.xml. Please check.

@rnveach @romani 
Please, make final review to make sure that we all agree on implementation and logic and I will generate diff reports.

romani commented:
> I cannot decide whether getAccessModifierFromModifiersToken should be in ChecksUtils or in AccessModifier.

please keep it in ChecksUtils, Enum should be simple a care only about itself. Enum should not know where it is used.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AccessModifier.java:
```diff
@@ -0,0 +1,63 @@
+    /** Protected access modifier. */
+    PROTECTED,
+    /** Package-local access modifier. */
+    PACKAGE_LOCAL,
```
From where you get "package local" term ?

I see only "package" and "package-private":
>or package-private (no explicit modifier).

at https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html

> If a class or interface type is declared with package access, then it may be accessed only from within the package in which it is declared. 

at https://docs.oracle.com/javase/specs/jls/se8/html/jls-6.html#jls-6.6.1

>  In the absence of an access modifier, most declarations have package access. A class or interface type declared without an access modifier implicitly has package access. 

at https://docs.oracle.com/javase/specs/jls/se8/html/jls-6.html
http://bugs.java.com/view_bug.do?bug_id=4456057 (Java representative in evaluation also call it "package" )

---------------

I would rather stay with JDK terminology - "package access".
"package-private" looks like is more human friendly term, I do not think we should use it.
@rnveach , please share your opinion.

rnveach replied:
@romani JDK terminology is fine with me.
Why not just name it just `PACKAGE`? adding `access` in an class named `access` seems redundant. If comment followed pattern it would read `Package-access access modifier`.

romani replied:
Name should be `PACKAGE` . I just qouted jdk to ease search on the pages.

MEZk replied:
@romani 
Done

rnveach commented on src/xdocs/config_naming.xml:
```diff
@@ -1185,15 +1185,15 @@ public boolean equals(Object o) {
-            <td><code>anoninner</code></td>
+            <td>accessModifiers</td>
+            <td>Access modifiers of methods where parameters are checked.</td>
+            <td><a href="property_types.html#access_modifier">Access Modifier Set</a></td>
```
Should be `access_modifiers` as detailed in `property_types.xml`.

MEZk replied:
Fixed

rnveach commented on src/xdocs/config_naming.xml:
```diff
@@ -1185,15 +1185,15 @@ public boolean equals(Object o) {
-            <td><a href="property_types.html#scope">Scope</a></td>
+            <td>excludeAccessModifiers</td>
+            <td>AccessModifiers of methods where parameters are not checked.</td>
+            <td><a href="property_types.html#access_modifier">Access Modifier Set</a></td>
```
Same here.

MEZk replied:
excludeAccessModifiers was removed

rnveach commented on src/xdocs/config_naming.xml:
```diff
@@ -1226,19 +1226,11 @@ public boolean equals(Object o) {
         </source>
         <p>
           The following configuration checks that the parameters always start with two
-          lowercase characters and, in addition, that public method parameters cannot be one character
```
it seems to me we are removing the original purpose of this example, which is how to configure this check differently between public and non-public naming conventions.
I think we should restore it. It should be similar to google config, one should be `public` alone and the other should have all the other modifiers.

MEZk replied:
Restored. Extended example.

rnveach commented on src/xdocs/property_types.xml:
```diff
@@ -403,6 +403,18 @@
+          <p>This property represents Java access modifiers.</p>
+
+          <ul>
+              <li><code>null</code></li>
```
I find adding `null`/`nothing` weird. Users cannot specifically set this value, right?
`Scope` has `nothing` but like I said, they can't specifically write a `nothing` value in the configuration.

Either one should throw an exception that it isn't in the enumeration if users try to use them.
My opinion is we just don't list these. Checks have default values, so that should be enough.

MEZk replied:
Fixed

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/ParameterNameCheck.java:
```diff
@@ -70,19 +71,19 @@
-    private Scope excludeScope;
+    /** Access modifiers of methods which should be checked. */
+    private AccessModifier[] accessModifiers =
+        Stream.of(AccessModifier.PUBLIC).toArray(AccessModifier[]::new);
```
Oirignal check looked at all modifers, not just `public`. I think we should continue this.
If we don't specify all, I think more examples in xdoc need to be changed/added.

MEZk replied:
Added public, package, private, protected modifiers into accessModifiers default values.

MEZk commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/CheckUtils.java:
```diff
@@ -425,4 +427,33 @@ public static boolean isReceiverParameter(DetailAST parameterDefAst) {
+        for (AST token = modifiersToken.getFirstChild(); token != null;
+             token = token.getNextSibling()) {
+
+            if ("public".equals(token.getText())) {
```
Fixed

MEZk commented on src/test/java/com/puppycrawl/tools/checkstyle/utils/CheckUtilsTest.java:
```diff
@@ -112,7 +113,34 @@ public void testEquals() {
+
+        try {
+            CheckUtils.getAccessModifierFromModifiersToken(modifiers);
+            fail(IllegalArgumentException.class.getSimpleName() + "was expcted.");
```
Fixed

MEZk commented on src/test/java/com/puppycrawl/tools/checkstyle/utils/CheckUtilsTest.java:
```diff
@@ -112,7 +113,34 @@ public void testEquals() {
+    public void testGetAccessModifierFromModifiersTokenWithNullParameter() {
+        try {
+            CheckUtils.getAccessModifierFromModifiersToken(null);
+            fail(IllegalArgumentException.class.getSimpleName() + "was expcted.");
```
Fixed

MEZk commented:
@rnveach 

>Oirignal check looked at all modifers, not just public. I think we should continue this.
If we don't specify all, I think more examples in xdoc need to be changed/added.

I generated diff reports and there are no differences with default configuration (when check validates all modifiers).

romani commented:
lets doc testing by two reports:
1) there should be diff as we do expect less violations that are reported in issue.
2) implementation of 7.4 was tested on guava code and we reviewed all cases, so reports from checkstyle 7.4 + google config of 7.4 == checkstyle 7.5 + google config 7.5

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/CheckUtils.java:
```diff
@@ -425,4 +427,33 @@ public static boolean isReceiverParameter(DetailAST parameterDefAst) {
+        for (AST token = modifiersToken.getFirstChild(); token != null;
+             token = token.getNextSibling()) {
+
+            if (AccessModifier.PUBLIC.getName().equals(token.getText())) {
```
I think you mis-understood what I meant. I meant token type like,
`if (TokenTypes.LITERAL_PUBLIC == token.getType()) {`
I see no reason to compare any type of text, when these tokens have standard integer types that already specify them.

MEZk replied:
Done

rnveach commented on src/xdocs/config_naming.xml:
```diff
@@ -1185,18 +1185,12 @@ public boolean equals(Object o) {
-            <td><code>null</code></td>
-          </tr>
-        </table>
+        </table>the
```
remove `the`.

MEZk replied:
Done

MEZk commented:
@romani 
Diff reports:
> 1. there should be diff as we do expect less violations that are reported in issue.

http://mezk.github.io/i3675-ParameterNameCheck/diff1/index.html

> 2. implementation of 7.4 was tested on guava code and we reviewed all cases, so reports from checkstyle 7.4 + google config of 7.4 == checkstyle 7.5 + google config 7.5

http://mezk.github.io/i3675-ParameterNameCheck/diff2/index.html

romani commented:
TC is not happy

MEZk commented:
@romani 

TC reports minors

>448: getAccessModifierFromModifiersToken() TokenTypes.LITERAL_PUBLIC == tokenType: constant on left side of comparison
451: getAccessModifierFromModifiersToken() TokenTypes.LITERAL_PROTECTED == tokenType: constant on left side of comparison
454: getAccessModifierFromModifiersToken() TokenTypes.LITERAL_PRIVATE == tokenType: constant on left side of comparison

I will fix them. Please, take a look at reports. There are many differences, but we can check at least MaterialDesignLibrary (208)
java-design-patterns (203)
protonpack (218)
sevntu-checkstyle-with-excludes (178)

MEZk commented:
@romani 
> 2. implementation of 7.4 was tested on guava code and we reviewed all cases, so reports from checkstyle 7.4 + google config of 7.4 == checkstyle 7.5 + google config 7.5

Config before fix:
```xml
        ]module name="ParameterName">
            <property name="id" value="ParameterNameNonPublic"/>
            <property name="format" value="^[a-z]([a-z0-9][a-zA-Z0-9]*)?$"/>
            <property name="excludeScope" value="public"/>
            <message key="name.invalidPattern"
             value="Parameter name ''{0}'' must match pattern ''{1}''."/>
        </module>
        <module name="ParameterName">
            <property name="id" value="ParameterNamePublic"/>
            <property name="format" value="^[a-z][a-z0-9][a-zA-Z0-9]*$"/>
            <property name="scope" value="public"/>
            <message key="name.invalidPattern"
             value="Parameter name ''{0}'' must match pattern ''{1}''."/>
        </module>
```

Config after fix:
```xml
        <module name="ParameterName">
            <property name="id" value="ParameterNameNonPublic"/>
            <property name="format" value="^[a-z]([a-z0-9][a-zA-Z0-9]*)?$"/>
            <property name="accessModifiers" value="protected, package, private"/>
            <message key="name.invalidPattern"
             value="Parameter name ''{0}'' must match pattern ''{1}''."/>
        </module>
        <module name="ParameterName">
            <property name="id" value="ParameterNamePublic"/>
            <property name="format" value="^[a-z][a-z0-9][a-zA-Z0-9]*$"/>
            <property name="accessModifiers" value="public"/>
            <message key="name.invalidPattern"
             value="Parameter name ''{0}'' must match pattern ''{1}''."/>
        </module>
```

Diff reports:
https://mezk.github.io/i3675-ParameterNameCheck/diff2/

Reports looks OK as for me.

romani commented:
All reports are fine.
